### PR TITLE
Update to NC14

### LINKED ISSUE #491: Compatibility with Nextcloud 14 [$190]
Hi @brantje,

as you may have already seen, we have entered the beta phase for the release of Nextcloud 14. We try to keep up compatibility with older apps, but in some cases apps need some adjustments to properly work on new Nextcloud releases. With Nextcloud 14 there have been quite some changes in that regard:

You can find a list of all critical changes for app developers and admins at https://github.com/nextcloud/server/issues/7827 You can run `occ app:check-code` to validate that there are no issues with deprecated or private message calls. 

One critical change is the overall restructuring of the main template (https://github.com/nextcloud/server/pull/9982) that might require some adjustments in your app.

We would be very pleased if you can make your app compatible with Nextcloud 14 and release a new version to the app store, so that users can properly test it and the upgrade experience to Nextcloud 14 will be even better than before.

Feel free to ask, if you have any questions regarding making your app compatible with Nextcloud 14. 

Thanks for being part of the Nextcloud community and for all the effort you put into providing this app :rocket:

Btw. you might have seen that starting August 25, the Nextcloud Conference takes place at the familiar TU Berlin. (https://nextcloud.com/conf/)! We'd like all contributors to be there and that most certainly includes you. You might already have registered but if you haven't - please consider joining us!

<bountysource-plugin>

---
There is a **[$190 open bounty](https://www.bountysource.com/issues/61972318-compatibility-with-nextcloud-14?utm_campaign=plugin&utm_content=tracker%2F44880056&utm_medium=issues&utm_source=github)** on this issue. Add to the bounty at [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F44880056&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by newhinton**
This is the pullrequest for the ui-fixes in nextcloud 14
closes #491

### CODE REVIEW

**Comment by newhinton** [comment]
@jospoortvliet @juliushaertl @vwbusguy @brantje @animalillo
This branch updates the passman app to nc14.

**Comment by joergmschulz** [comment]
just unzipped this branch to my NC14 - the view doesn't seem to be perfect: 
grafik
(this is the left pane)

**Comment by newhinton** [comment]
@joergmschulz Did you run 'grunt watch' or 'grunt build' ? the branch currently only contains the sources, and is not releaseready

**Comment by joergmschulz** [comment]
> @joergmschulz Did you run 'grunt watch' or 'grunt build' ? the branch currently only contains the sources, and is not a full release. The generated files by the build process are likely missing

no, I didn't, and: yes, npm install  ; grunt build helped

**Comment by newhinton** [comment]
Great news! have fun with it :D

**Comment by MagicFab** [comment]
The website and demo are down as well:
* https://demo.passman.cc/
* https://passman.cc/

**Comment by earth08** [comment]
Dear,
How to get your patch to work with nextcloud 14,

@brantje is sleeping,
May be

**Comment by newhinton** [comment]
@earth08 have you run 'grunt build'?

**Comment by earth08** [comment]
Sorry dear,
I don't know how to use gruntbuild,
Can u help me

**Comment by newhinton** [comment]
you need the following packages:
nodejs and npm

install them in ubuntu with 'sudo apt-get install nodejs npm'

after that, switch to your passman directory

run 'npm install grunt'
and after that
'grunt build'

that should fix the problems

**Comment by newhinton** [comment]
@e-alfred Could you explain why this is important for this commit? While it is deprecated, it is still in the code, so passman should be fine for now. Does this break anything for you?

**Comment by e-alfred** [comment]
@newhinton I think it is better to fix things now than later on. This way it is one less problem if Nextcloud 15 is released and compatibility problems arise.

**Comment by RCAPDART** [comment]
> @newhinton I think it is better to fix things now than later on. This way it is one less problem if Nextcloud 15 is released and compatibility problems arise.

I agree, that deprecated items should be replaced by actual ones, but this fix will be out of the current scope. And it will slow down process of this PR (which is critical now).
One ticket = one PR.

**Comment by newhinton** [comment]
@RCAPDART Thank you! I definitively will take a look at you suggestions there, it's always good to have such help!


@e-alfred You are right, such deprecated methods should be removed in the near future. But i have to agree with @RCAPDART, i don't think they belong here, but i will also take a look at this on the weekend, and create a fix for this in a different PR. So it should be fixed in a future release, just not the next one ;)

**Comment by beanaroo** [comment]
Thank you, @newhinton for your dedication. I've been following the NC14 status.

I noticed fontawesome resources are not loading and had to move them up from `/css/vendor/font-awesome/` into `/css`. I'm not sure if that's expected or if didn't build properly. Just wanted to mention.

**Comment by newhinton** [comment]
The problem with the overlapping sidebar and the missing tags is now solved, everything is beeing displayed just fine

File: appinfo/info.xml [hunk 1]
```diff
@@ -19,7 +19,7 @@ Features:
 - Import from various password managers (KeePass, LastPass, DashLane, ZOHO, Clipperz.is )
 For an demo of this app visit [https://demo.passman.cc](https://demo.passman.cc)
 ]]></description>
-    <version>2.1.5</version>
+    <version>2.2.0</version>
```
Comment (animalillo) [inline appinfo/info.xml]: I think that @brantje wanted to keep 2.0 for when sharing vaults was ready, can you confirm @brantje?

Comment (newhinton) [reply appinfo/info.xml]: @animalillo I thought this update would exceed the patch in your MAJOR.MINOR.PATCH versioning, but if you disagree with that, it'll be 2.1.6
just say a word

**Comment by animalillo** [comment]
I remember this menu centered on the screen and maybe a bit wider
imagen

Deleted credential icon overlaps with the text and the button with the settings menu
imagen
imagen

Tag icon overlaps with tag text
imagen

Credential list have random icon size breaking table line consistency
imagen

**Comment by brantje** [comment]
image
Spinner should be centered

image
Add button should be next to the folder tree.
like this:
image

image

**Comment by animalillo** [comment]
Now it looks better. I have found a few issues though.

 - [ ] When i try to interact with the quick menu (side menu when you click on a credential on the list) if i try to make a password visible the menu instantly hides, same for copy or other actions.

 - [ ] The revisions list still looks a bit broken.
imagen

 - [ ] I got lost for almost a minute struggling to find the create credential button till i noticed that after my vault name it's a random "New" text that seems to open the credential creation menu. If I found trouble finding this and people used to have trouble with the old button style already i don't know if an average user will be able to find it, I'm ok with a bit of redesign on that part, but can you think on another way to make it more visible instead of less visible?
imagen


 - [ ] Share menu also looks broken, probably you missed this and the revisions one
imagen
imagen

 - [ ] Also, the donate button is not doing anything.

**Comment by animalillo** [comment]
Did you check our test protocol?
https://github.com/nextcloud/passman/wiki/Test-Protocol

**Comment by brantje** [comment]
Revisions look weird
image

 Changing icon's doesn't work

**Comment by newhinton** [comment]
@brantje @animalillo The issues should be fixed.

Icon-changing does work for me, and if it doesn't, thats not really part of this pr. Please open an issue for that seperately.

Oh i see you edited your comment regarding icons: i guess thats related to #486

**Comment by earth08** [comment]
i noticed two things which needs attention to fix
1) when we check for password audit and when we find some poor password, actions link doesnt work
2) when sharing password with link the layout to add email and other things on right side doesnt seems to work
screenshot from 2018-10-17 22-27-40

**Comment by earth08** [comment]
Dear thanks for providing update
but action link doesnt work,
and when we open any of our saved password from password list and after opening sidebar we cannot see the password, (when clicking on any of items other than edit, it takes back me to password lists
one thing i noted today on my minimized chrome, the main page and login page get cut,
screenshot from 2018-10-27 21-11-04
screenshot from 2018-10-27 21-11-11
Please check this

**Comment by newhinton** [comment]
@earth08 

Changes not being visible until a relogin is a different Issue, #486.
I have created an hotfix for now, but until i found out what the reason is for this behaviour, only changes to existing passwords will be visible without reloging.

Are you sure that the actionlink still doesn't work? this should be fixed in one of the latest commits

I will take a look at the vaultlist-site

Also the off-click in the edit section should be fixed, but ill take a look at it

Thanks for your review!

**Comment by earth08** [comment]
Dear, thanks for your hard work for the community,
this is password audit page
screenshot from 2018-10-28 09-16-52
when i check for passwords, it shows me 9 weak passwords but it shows only one.
one error here is, 
when we change number of minimum password strength and check the earlier ones are not getting cleared, and the new list comes at top.
the main login screen for passman is not align and still cuts when browser is half the screen or smaller.

one more thing to look at, when we change or add tag to any item it reflects after relogin, means there are many things which gets reflected after relogin

**Comment by newhinton** [comment]
@earth08 Have you tried running 'grunt build'?

**Comment by animalillo** [comment]
why did you rewrite all this js files if the issues are mere visual fixes?

imagen

**Comment by animalillo** [comment]
Why did you change tags behavior?
This merge request should not change functionality regarding passman behavior, that should go in other merge request and you should ask before doing those changes to check if they make sense.

For us, and a lot of people using tags, the search feature in the tags section is a must, since people have a lot of tags on their passwords as the vault grows with time. I have hundreds of tags and using the search box is a life savior.

I like the idea of a feature to filter passwords by the magic tags you made, but not if that must kill the filter tags. For now I would focus this merge request on making passman work as it used to more than adding new features. But don't dump the idea, just take it out for now and let's talk on telegram on how to improve it! :D

this
imagen

should look more like this
imagen

**Comment by earth08** [comment]
I cannot use grunt build on my arm device,
It throws me door of task sass not available

**Comment by newhinton** [comment]
@earth08 did you run 'apt install ruby sass' or the corresponding packagemanager for your distribution?

**Comment by jospoortvliet** [comment]
So it seems @brantje is entirely inactive. If this branch works, @newhinton - I'd be happy to have you take over ownership. Now this PR is getting very big, @animalillo has a point that changes of behavior probably would be better off in a separate PR but either case - you're active, how about you run the show? Get this merged if it works and get a release out... 14 is still not supported and 15 is around the corner ;-)

**Comment by jospoortvliet** [comment]
I don't know exactly how releases work but I'm sure our sysadmins can help you update the existing app, if @brantje doesn't object in the coming days.

**Comment by brantje** [comment]
@jospoortvliet Ik ben er nog wel hoor. Alleen niet zo actief op github.
Again i do not give permission to take over this repo.

**Comment by newhinton** [comment]
@jospoortvliet I am in contact with @brantje, at least i was a few days ago. We are mainly in contact over the telegram group, so i guess it flies under the radar

**Comment by brantje** [comment]
@jospoortvliet Please transfer the repo back to my personal namespace. I've had it with the nextcloud attitude towards @animalillo and me.

**Comment by jospoortvliet** [comment]
@brantje Oh, sorry, I thought you had no more interest in maintain the app - no intention to insult, it's just that ppl keep asking about a release and your github history shows no commits here for a long time. Not sure what attitude you have a problem with, tbh, I simply want to make sure apps that go unmaintained are handed over to others that are active. But if you want the repo moved over, that's good with me - I don't know how to do it so we'll have to ask Morris or another dev to help out with that.

**Comment by brantje** [comment]
Me not being active is not a reason to give admin permissions to another user (nothing against @newhinton), without asking me.
I ask again, nicely to transfer the repo back to my namespace, i can't do it since i'm not an org owner, once the repo is transfered i'll make @newhinton admin, this time it is with my approval.

**Comment by animalillo** [comment]
I have been following this merge request and I think that @jospoortvliet should read the full thread on this merge request to find @brantje comments on it not too long ago.

@brantje has been extremely busy lately but I have been able to grasp some extra time to review the merge request and share the needs of it around here and around telegram a lot, and I hadn't been asked a thing about this  either.

Nextcloud should not ask for repos and say that their owners have full control of them if they don't plan to keep their word. Specially on security related products.

**Comment by animalillo** [comment]
Nextcloud some how seems to have a lot of interest on this project but none on the developers that created it nor any respect for them.

If nextcoud checked the inbox of job applications more often, they would have seen my job application and could have hire me to work full time on this project or anything else they've wanted.

I have applied half a month ago or so for a nextcloud backend developer job and didn't even get an answer back to my email, not even a "we will check your profile" boring answer. I understand that nextcloud might be busy or not interested on me, but it makes it seem like they also have no interest on programmers anymore, who knows.

I can take being ignored at a job post but I won't accept aggressive actions on opensource projects that have taken a lot of work from me and @brantje

**Comment by animalillo** [comment]
We would also like to have passman-android and passman-webextension transferred back please.

**Comment by jospoortvliet** [comment]
I think you guys are exaggerating quite a bit. I haven't seen a release for an eternity, people regularly ask on twitter if the project is dead - and I see @newhinton active yet nothing merged. I admit, I didn't spend a lot of time looking if somewhere, anywhere, there was a sign of life from Sander before I offered to give Newhinton the ability to do more than push commits in an ever growing PR that looked like it would never get released. Sorry for caring about the app - I feel like I shouldn't have bothered with it, clearly trying to help pisses everyone here off.

Doing nothing seems far better, so that's what I'll do going forward -> ignore.

**Comment by newhinton** [comment]
@Luticus Thank you for your review! I will take a look at your findings, and fix them :)

**Comment by Luticus** [comment]
That's awesome, I'll keep pulling your code down as I notice changes and test things out for ya on my server.  Thanks for all the work you're doing!  Can't wait for everything to be merged.  :)

**Comment by brantje** [comment]
@nextcloud/community-mgmt 
Can you please move the following repo's
passman-webextension -> @brantje 
passman -> @brantje 
passman-android -> @animalillo

**Comment by Luticus** [comment]
Something else you guys might want to take a look at:  When you click on a credential in a vault and open the sidebar, in there if you click the eye icon (toggle visibility) the side bar will simply close instead of revealing the password as it should.

Also, let me know if you'd rather I make individual issues in the issue tracker when I find something.  I just felt like this would be a good place to put stuff that seemed to relate to the NC14 upgrade.

**Comment by newhinton** [comment]
@Luticus This should already be fixed, at least i cant reproduce it

**Comment by Luticus** [comment]
@newhinton  That's odd.  I tested it on chrome and firefox both with both my main nextcloud 14 server and my test nextcloud 14 server, which has almost no extra apps and I see it do it on all instances.  In addition my wife also said she sees it on her set up as well.  This is with a fresh pull from the fix_layout_for_14 branch.  So unless you've got newer code that hasn't been pushed yet, or maybe I didn't describe it well enough and you're looking at the wrong thing or something, I'm not sure what's up.  :)  Just to make sure we're on the same page I've attached a screenshot:
passman_bug

**Comment by newhinton** [comment]
@Luticus I can only reproduce the error if i don't run 'grunt build'. This however is nessessary to create all the compiled files for the correct operation of passman

**Comment by animalillo** [comment]
Hello @Luticus 
Why are you running grunt build as root? D:
It will mess your permissions on the generated files, you should never run such commands as root as a rule of thumb.
Now you should check file permissions so that your web server is the owner of the files.

Aditionally try clearing cache (ctrl+f5 should also do the trick) to get the fresh files in case your browser is caching the old version.

Hope this helps!

**Comment by Luticus** [comment]
By the way, now that the grunt build is working for me... the latest version of the app looks really nice!  I'll keep testing it out.

**Comment by e-alfred** [comment]
@newhinton I tested your PR on both Nextcloud 14 and 15 RC2 and it works pretty well. There is only one problem I noticed: If I save a new password, the credentials list doesn't get updated. I have to refresh the page and log in again to see it in the list.

**Comment by newhinton** [comment]
@e-alfred See #515

File: css/app.css [hunk 2]
```diff
@@ -126,67 +119,6 @@
...
- * This program is free software: you can redistribute it and/or modify
- * it under the terms of the GNU Affero General Public License as
- * published by the Free Software Foundation, either version 3 of the
- * License, or (at your option) any later version.
- *
- * This program is distributed in the hope that it will be useful,
- * but WITHOUT ANY WARRANTY; without even the implied warranty of
- * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
- * GNU Affero General Public License for more details.
- *
```
Comment (animalillo) [inline css/app.css]: Why did you remove the coppyright header?

**Comment by Luticus** [comment]
Since your last changes the list of tags don't seem to close for me.  Could be that I just updated to Nextcloud 15 too.  In any case, I thought I'd point it out.  Seems to work and look great on NC15 though, despite that one minor issue.

**Comment by newhinton** [comment]
@Luticus Did you run grunt build again? It may be that the minified javascript didn't get updated properly, can you try that? On my instances its working fine

**Comment by newhinton** [comment]
Hey guys! We are almost ready to publish passman 2.2.0 for Nextcloud 14 and 15! The Update includes many fixes to bugs reported by you, and i also overhauled the userinterface! Stay tuned for the update in the close future!

**Comment by Luticus** [comment]
I did run grunt build, I'll rerun it just in case.  It's probably an issue on my end.  It's pretty obvious so it's likely something you'd easily notice if it were reproducible.  I'll check into it again.

Also, a HUGE **thank you** for fixing this awesome app!  I'm stoked about the release and super happy it works on NC15 too.

**Comment by Luticus** [comment]
Yea a re-run of grunt didn't seem to fix it.  I also upgraded to php 7.1 from php7.0 so maybe that, although I suspect it would be JS and not php that would control the collapse feature of that menu.  Oh well, I suspect once the release happens it'll be a non-issue anyway and everything else seems to work fine.  I can wait.

**Comment by newhinton** [comment]
@Luticius In that case, i wi treat this as not reproducible, but IF you encounter any issues with the full release, please open an ticket for me and i will have another look ;)

**Comment by Luticus** [comment]
@newhinton Will do, thanks again!

**Comment by earth08** [comment]
Very excited and waiting eagerly for new version and excellent ui

**Comment by j0nnymoe** [comment]
@newhinton Any ETA on when this will be released? Thanks
